client/executor/wasmtime/src/instance_wrapper.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Defines data and logic needed for interaction with an WebAssembly instance of a substrate
+//! runtime module.
+
+use crate::util;
+use sc_executor_common::error::{Error, Result};
+use sp_wasm_interface::{Pointer, WordSize};
+use std::slice;
+use wasmtime::{Instance, Memory, Table};
+
+/// Wrap the given WebAssembly Instance of a wasm module with Substrate-runtime.
+///
+/// This struct is a handy wrapper around a wasmtime `Instance` that provides substrate specific
+/// routines.
+pub struct InstanceWrapper {
+	instance: Instance,
+	memory: Memory,
+	table: Option<Table>,
+}
+
+impl InstanceWrapper {
+	/// Wrap the given `Instance`.
+	///
+	/// # Safety
+	///
+	/// This wrapper requires exclusive ownership over the given instance and all its components.
+	/// Since `Instance` is a `Clone` we cannot use typesystem to prevent this, hence unsafe.
```

Comment:
We can (and should) use `RefCell` or `Mutex` for this.

Reply:
I am not sure if I agree: how would they help from violating the invariant. It would still be possible to clone the memory instance from within this module and create aliases, right?